File: src/gaze.py
from landmarks import RIGHT_EYE, RIGHT_IRIS, LEFT_EYE, LEFT_IRIS


def get_landmark_xy(face_landmarks, index):
    landmark = face_landmarks[index]

    return landmark.x, landmark.y


def get_eye_bounds(face_landmarks, eye_indices):
    xs = []
    ys = []

    for index in eye_indices:
        x, y = get_landmark_xy(face_landmarks, index)

        xs.append(x)
        ys.append(y)

    return (
        min(xs),
        max(xs),
        min(ys),
        max(ys)
    )


def get_iris_center(face_landmarks, iris_indices):
    xs = []
    ys = []

    for index in iris_indices:
        x, y = get_landmark_xy(face_landmarks, index)

        xs.append(x)
        ys.append(y)

    return (
        sum(xs) / len(xs),
        sum(ys) / len(ys)
    )


def normalize_iris_position(
    iris_x,
    iris_y,
    eye_bounds
):
    eye_left, eye_right, eye_top, eye_bottom = eye_bounds

    eye_width = eye_right - eye_left
    eye_height = eye_bottom - eye_top

    if eye_width == 0 or eye_height == 0:
        return 0.5, 0.5

    gaze_x = (iris_x - eye_left) / eye_width
    gaze_y = (iris_y - eye_top) / eye_height

    return gaze_x, gaze_y

class GazeSmoother:

    def __init__(self, alpha=0.2):
        self.alpha = alpha

        self.x = None
        self.y = None

    def update(self, x, y):

        if self.x is None:
            self.x = x
            self.y = y

        else:
            self.x = (
                self.alpha * x
                + (1 - self.alpha) * self.x
            )

            self.y = (
                self.alpha * y
                + (1 - self.alpha) * self.y
            )

        return self.x, self.y


def compute_gaze(face_landmarks, smoother):
    right_eye_bounds = get_eye_bounds(face_landmarks, RIGHT_EYE)
    left_eye_bounds = get_eye_bounds(face_landmarks, LEFT_EYE)

    right_iris_x, right_iris_y = get_iris_center(face_landmarks, RIGHT_IRIS)
    left_iris_x, left_iris_y = get_iris_center(face_landmarks, LEFT_IRIS)

    right_gaze_x, right_gaze_y = normalize_iris_position(
        right_iris_x, right_iris_y, right_eye_bounds
    )
    left_gaze_x, left_gaze_y = normalize_iris_position(
        left_iris_x, left_iris_y, left_eye_bounds
    )

    gaze_x = (right_gaze_x + left_gaze_x) / 2
    gaze_y = (right_gaze_y + left_gaze_y) / 2

    return smoother.update(gaze_x, gaze_y)

File: web/frontend/js/vision/gaze.js
/*
  src/gaze.py 의 JS 포팅.

  눈 bounding box -> 홍채 중심 -> 0~1 정규화 -> 좌우 평균 -> EMA 스무딩.
  전부 순수 산술이라 파이썬판과 동일한 입력에 동일한 출력이 나와야 한다
  (AI 없음, 상태는 GazeSmoother 의 EMA 뿐).

  ※ 합산 순서까지 파이썬과 맞춰놨다. 부동소수점 덧셈은 순서에 따라 마지막
    자리가 달라질 수 있어서, 수치 일치 검증을 통과시키려면 순서를 지켜야 한다.
*/

import { RIGHT_EYE, RIGHT_IRIS, LEFT_EYE, LEFT_IRIS } from "./landmarks.js";

export function getLandmarkXY(faceLandmarks, index) {
  const landmark = faceLandmarks[index];
  return [landmark.x, landmark.y];
}

export function getEyeBounds(faceLandmarks, eyeIndices) {
  const xs = [];
  const ys = [];

  for (const index of eyeIndices) {
    const [x, y] = getLandmarkXY(faceLandmarks, index);
    xs.push(x);
    ys.push(y);
  }

  return [Math.min(...xs), Math.max(...xs), Math.min(...ys), Math.max(...ys)];
}

export function getIrisCenter(faceLandmarks, irisIndices) {
  const xs = [];
  const ys = [];

  for (const index of irisIndices) {
    const [x, y] = getLandmarkXY(faceLandmarks, index);
    xs.push(x);
    ys.push(y);
  }

  let sumX = 0;
  let sumY = 0;
  for (const x of xs) sumX += x;
  for (const y of ys) sumY += y;

  return [sumX / xs.length, sumY / ys.length];
}

export function normalizeIrisPosition(irisX, irisY, eyeBounds) {
  const [eyeLeft, eyeRight, eyeTop, eyeBottom] = eyeBounds;

  const eyeWidth = eyeRight - eyeLeft;
  const eyeHeight = eyeBottom - eyeTop;

  if (eyeWidth === 0 || eyeHeight === 0) {
    return [0.5, 0.5];
  }

  return [(irisX - eyeLeft) / eyeWidth, (irisY - eyeTop) / eyeHeight];
}

export class GazeSmoother {
  constructor(alpha = 0.2) {
    this.alpha = alpha;
    this.x = null;
    this.y = null;
  }

  update(x, y) {
    if (this.x === null) {
      this.x = x;
      this.y = y;
    } else {
      this.x = this.alpha * x + (1 - this.alpha) * this.x;
      this.y = this.alpha * y + (1 - this.alpha) * this.y;
    }

    return [this.x, this.y];
  }
}

export function computeGaze(faceLandmarks, smoother) {
  const rightEyeBounds = getEyeBounds(faceLandmarks, RIGHT_EYE);
  const leftEyeBounds = getEyeBounds(faceLandmarks, LEFT_EYE);

  const [rightIrisX, rightIrisY] = getIrisCenter(faceLandmarks, RIGHT_IRIS);
  const [leftIrisX, leftIrisY] = getIrisCenter(faceLandmarks, LEFT_IRIS);

  const [rightGazeX, rightGazeY] = normalizeIrisPosition(rightIrisX, rightIrisY, rightEyeBounds);
  const [leftGazeX, leftGazeY] = normalizeIrisPosition(leftIrisX, leftIrisY, leftEyeBounds);

  const gazeX = (rightGazeX + leftGazeX) / 2;
  const gazeY = (rightGazeY + leftGazeY) / 2;

  return smoother.update(gazeX, gazeY);
}
